### ISSUE TITLE
Polymorphic Model

### ISSUE THREAD

**Issue by rmb938** [schematics/schematics#448]
I have a use case that requires polymorphic models. I am currently manually parsing my dictionaries to figure out what model they belong to but I would like if that was built into schematics.

See the following models:

```
kind: NGINX
port_range:
  - 8080
  - 8180
location_defaults:
  proxy_pass: "http://{{ location }}/"
  proxy_set_header:
    - "Host $host:$server_port"
    - "X-Real-IP $remote_addr"
    - "X-Forwarded-For $proxy_add_x_forwarded_for"
    - "X-Forwarded-Proto $scheme"
    - "X-Forwarded-Port {{ server_port }}"
  add_header: "X-Upstream $upstream_addr"
  rewrite: "/{{ location }}/(.*) /$1 break"
```

```
kind: Definition
name: grafana
image: "{{ '{{ docker_repos[\'default\'].repo }}' }}/cpe/grafana:latest"
nginx:
  grafana:
    port: 3000
service:
  grafana:
    port: 3000
volumes:
  - {{ volumes }}/grafana/grafana.ini:/etc/grafana/grafana.ini
```

I would like to be able to key off of 'kind' and have schematics figure out what model it actually is.

I found this comment https://github.com/schematics/schematics/issues/184#issuecomment-65777372 and would like something similar if possible.

**Comment by bintoro** [comment]
In the end, the thread you linked to resulted in `types.compound.PolyModelType` that can do precisely that. But it's a special case that will be deprecated eventually.

Today we also have `UnionType` which allows a field to accept one of a number of types.

What I want to do is
* improve `UnionType` so that the underlying types can include multiple `ModelType` instances
* give regular ol' `ModelType` the ability to do subclass detection à la `PolyModelType`.

Once that work is done, a separate `PolyModelType` is no longer needed.
